docs(retain): document context as caller-supplied trust hint + untrusted-content bank separation (#3559)
vectorize-io/hindsight: hindsight-docs/docs/developer/retain.md

**Note:** Observations are consolidated automatically in the background after `retain()` operations complete. This consolidation process synthesizes patterns from new facts into the bank's knowledge base.

### Trust boundaries (security note)

The `context` string is **caller-supplied and unauthenticated** — a
disambiguation hint, not a trust assertion. Any caller who can invoke
`retain()` can supply a misleading `context` (for example
`context="security_review_note"`) for content that did not come from that
source. Treat `context` as advisory only; do not rely on it as evidence of
provenance.

For content from untrusted sources (web scraping, third-party integrations, or
user-supplied documents), retain it into a **separate bank** and merge
selectively at `recall()`/`reflect()` time. Bank isolation is strict, so a

> dedicated untrusted bank is the reliable trust boundary rather than the

Reviewer: One precision fix: `recall()` and `reflect()` both target a single `bank_id` — there is no native cross-bank merge, so "merge selectively at `recall()`/`reflect()` time" could imply a capability Hindsight doesn't have. Because bank isolation is strict, `reflect()` on one bank can't see another bank's data; the reliable pattern is to query the untrusted bank separately and combine results in your own decision logic:

```suggestion
user-supplied documents), retain it into a **separate bank**. `recall()` and
`reflect()` target a single bank, so query the untrusted bank separately and
combine results in your own decision logic. Bank isolation is strict, so a
```

`context` label.

See [Memory Defense](#memory-defense-and-source-provenance) for the built-in
prompt-injection detection that runs at retain time.

> 

Reviewer: Great note — technically accurate and well placed. One addition that closes the loop: the doc already has a *real* provenance mechanism, `receipt_uri` (under **Memory Defense and Source Provenance**). Pointing readers at it turns this from a warning into a warning + solution:

```suggestion
See [Memory Defense and Source Provenance](#memory-defense-and-source-provenance)
for the built-in prompt-injection detection that runs at retain time; use
`receipt_uri` to record actual provenance (an external receipt or co-signature
pointer) rather than relying on `context`.
```

**Note:** Observations are consolidated automatically in the background after `retain()` operations complete. This consolidation process synthesizes patterns from new facts into the bank's knowledge base.

> ### Trust boundaries (security note)
> 
> The `context` string is **caller-supplied and unauthenticated** — a
> disambiguation hint, not a trust assertion. Any caller who can invoke
> `retain()` can supply a misleading `context` (for example
> `context="security_review_note"`) for content that did not come from that
> source. Treat `context` as advisory only; do not rely on it as evidence of
> provenance.
> 
> For content from untrusted sources (web scraping, third-party integrations, or
> user-supplied documents), retain it into a **separate bank** and merge
> user-supplied documents), retain it into a **separate bank**. `recall()` and
> `reflect()` target a single bank, so query the untrusted bank separately and
> combine results in your own decision logic. Bank isolation is strict, so a
> `context` label.
> 
> See [Memory Defense](#memory-defense-and-source-provenance) for the built-in
> See [Memory Defense and Source Provenance](#memory-defense-and-source-provenance)
> for the built-in prompt-injection detection that runs at retain time; use
> `receipt_uri` to record actual provenance (an external receipt or co-signature
> pointer) rather than relying on `context`.

Reviewer: ⚠️ Heads-up: applying both suggestions in one batch produced a garbled merge — the old `and merge` line wasn't removed and the Memory Defense link got duplicated. Here's the corrected section as a single suggestion (one click fixes it):

```suggestion
### Trust boundaries (security note)

The `context` string is **caller-supplied and unauthenticated** — a
disambiguation hint, not a trust assertion. Any caller who can invoke
`retain()` can supply a misleading `context` (for example
`context="security_review_note"`) for content that did not come from that
source. Treat `context` as advisory only; do not rely on it as evidence of
provenance.

For content from untrusted sources (web scraping, third-party integrations, or
user-supplied documents), retain it into a **separate bank**. `recall()` and
`reflect()` target a single bank, so query the untrusted bank separately and
combine results in your own decision logic. Bank isolation is strict, so a
dedicated untrusted bank is the reliable trust boundary rather than the
`context` label.

See [Memory Defense and Source Provenance](#memory-defense-and-source-provenance)
for the built-in prompt-injection detection that runs at retain time; use
`receipt_uri` to record actual provenance (an external receipt or co-signature
pointer) rather than relying on `context`.

```
